File: apps/Applications/ios/Local For Vocal/Local For Vocal/Sources/Views/Pages/GroceryPageView.swift
import SwiftUI

struct GroceryPageView: View {
    @Binding var activeTab: TabType
    @State private var searchText = ""
    @EnvironmentObject var basketManager: BasketManager

    // For typing animation
    @State private var typingText = ""
    private let fullText = "Lowest price..."
    @State private var isDeleting = false

    @State private var isSearching = false
    @State private var scrollOffset: CGFloat = 0
    @State private var refreshID = UUID()

    var body: some View {
        ZStack(alignment: .top) {
            // ── Yellow-to-White Gradient Background ──
            LinearGradient(
                gradient: Gradient(colors: [
                    Color(hex: "#FFF8E7"),  // Warm yellow top
                    Color(hex: "#FFFDF5"),  // Light transition
                    Color.white,  // White bottom
                ]),
                startPoint: .top,
                endPoint: .bottom
            )
            .ignoresSafeArea()

            ScrollView {
                LazyVStack(spacing: 0, pinnedViews: [.sectionHeaders]) {
                    // ── Scrollable Header (tabs + location) ──
                    // This scrolls away with content
                    GroceryStaticHeader(activeTab: $activeTab)

                    Section(
                        header:
                            VStack(spacing: 0) {
                                GroceryStickyHeader(
                                    text: $searchText,
                                    onSearchTap: {
                                        isSearching = true
                                    }
                                )
                            }
                            .frame(maxWidth: .infinity)
                            .background(Color(hex: "#FFF8E7"))
                            .ignoresSafeArea(edges: .top)
                            .zIndex(100)
                            .shadow(color: Color.black.opacity(0.05), radius: 3, x: 0, y: 2)
                    ) {
                        // Main Content Driven by SDUI
                        VStack(spacing: 0) {
                            SDUIPage(slug: "grocery")
                                .id(refreshID)  // Force recreation on refresh
                        }
                        .padding(.top, 16)
                        .padding(.bottom, 100)
                        .zIndex(-1)
                    }
                }
            }
            .refreshable {
                // Hard refresh: Clear cache and force reload
                await SDUICacheManager.shared.invalidate(slug: "grocery")
                try? await Task.sleep(nanoseconds: 500_000_000)  // Small delay for UX
                await MainActor.run {
                    refreshID = UUID()
                }
            }

            // ── Fixed Status Bar Background ──
            Color(hex: "#FFF8E7")
                .ignoresSafeArea(edges: .top)
                .frame(height: 0)  // Allows it to extend only into the safe area
                .zIndex(200)
        }
        .onAppear {
            startTypingAnimation()
        }
        .fullScreenCover(isPresented: $isSearching) {
            GroceryGlobalSearchView()
        }
    }

    func startTypingAnimation() {
        Timer.scheduledTimer(withTimeInterval: 0.15, repeats: true) { timer in
            if !isDeleting {
                if typingText.count < fullText.count {
                    let index = fullText.index(fullText.startIndex, offsetBy: typingText.count)
                    typingText.append(fullText[index])
                } else {
                    // Wait before deleting
                    DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) {
                        isDeleting = true
                    }
                }
            } else {
                if !typingText.isEmpty {
                    typingText.removeLast()
                } else {
                    isDeleting = false
                }
            }
        }
    }
}

// MARK: - Subcomponents

struct GroceryStaticHeader: View {
    @Binding var activeTab: TabType

    var body: some View {
        VStack(spacing: 0) {
            // Top Tab Switcher
            TopCategoryBoxesView(activeTab: $activeTab)

            HStack {
                HStack(spacing: 8) {
                    Image(systemName: "house.fill")
                        .foregroundColor(Color(hex: "#8B6914"))
                        .font(.system(size: 16))

                    Text("WORK")
                        .font(.system(size: 11, weight: .bold))
                        .foregroundColor(Color(hex: "#8B6914"))

                    Text("Select your location")
                        .font(.system(size: 12))
                        .foregroundColor(Color(hex: "#6B5720"))
                        .lineLimit(1)

                    Image(systemName: "chevron.right")
                        .foregroundColor(Color(hex: "#8B6914"))
                        .font(.system(size: 14))
                }

                Spacer()

                HStack(spacing: 4) {
                    Text("\(Date().formatted(.dateTime.day()))")
                        .font(.system(size: 12, weight: .bold))
                        .foregroundColor(.white)
                    Text(Date().formatted(.dateTime.month()))
                        .font(.system(size: 12, weight: .bold))
                        .foregroundColor(.white)
                    Text(Date().formatted(.dateTime.weekday()))
                        .font(.system(size: 9))
                        .foregroundColor(.white.opacity(0.9))
                }
                .padding(.horizontal, 12)
                .padding(.vertical, 6)
                .background(Color(hex: "#FF9800"))
                .cornerRadius(8)
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 8)
            .background(Color(hex: "#FFF8E7"))
        }
    }
}

struct GroceryStickyHeader: View {
    @Binding var text: String
    var onSearchTap: () -> Void

    var body: some View {
        HStack(spacing: 12) {
            Button(action: onSearchTap) {
                HStack {
                    Image(systemName: "magnifyingglass")
                        .foregroundColor(Color(hex: "#9CA3AF"))
                    Text("Search for atta, dal, oil...")
                        .font(.system(size: 14))
                        .foregroundColor(Color(hex: "#6B7280"))  // Placeholder gray
                    Spacer()
                }
                .padding(.horizontal, 14)
                .padding(.vertical, 12)
                .background(Color.white)
                .cornerRadius(10)
                .overlay(
                    RoundedRectangle(cornerRadius: 10)
                        .stroke(Color(hex: "#E5E7EB"), lineWidth: 1)
                )
                .shadow(color: Color.black.opacity(0.05), radius: 2, x: 0, y: 1)
            }
            .buttonStyle(PlainButtonStyle())

            Button(action: onSearchTap) {  // Mic also triggers search
                Image(systemName: "mic.fill")
                    .foregroundColor(Color(hex: "#FF9800"))
                    .frame(width: 46, height: 46)
                    .background(Color.white)
                    .cornerRadius(10)
                    .overlay(
                        RoundedRectangle(cornerRadius: 10)
                            .stroke(Color(hex: "#E5E7EB"), lineWidth: 1)
                    )
            }
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 8)
        .background(Color(hex: "#FFF8E7"))
    }
}

struct GroceryHeroBanner: View {
    let typingText: String

    var body: some View {
        ZStack {
            // Background Image/Color
            AsyncImage(
                url: URL(
                    string:
                        "https://res.cloudinary.com/deljcbcvu/image/upload/v1768337045/Grocery_Offer_Backgroung_krgtp0.jpg"
                )
            ) { image in
                image.resizable()
                    .aspectRatio(contentMode: .fill)
            } placeholder: {
                Color(hex: "#FF9800")
            }
            .frame(minHeight: 280)
            .cornerRadius(16)
            .clipped()

            VStack(spacing: 20) {
                // Typing Text
                VStack(spacing: 0) {
                    Text(typingText + "|")
                        .font(.custom("Snell Roundhand", size: 38))  // Or system serif/cursive
                        .fontWeight(.bold)
                        .foregroundColor(.white)
                        .shadow(color: Color.black.opacity(0.4), radius: 4, x: 1, y: 1)
                }
                .frame(height: 55)

                // Categories
                HStack(spacing: 12) {
                    GrocerySubCategoryCard(
                        title: "Winter Essential",
                        discount: "Up to 70% off",
                        color: "#1B5E20",
                        iconUrl:
                            "https://res.cloudinary.com/deljcbcvu/image/upload/v1768339003/products/z9nxbc93hgfkkogpflbk.png"
                    )
                    GrocerySubCategoryCard(
                        title: "Snack & Sip",
                        discount: "Up to 50% off",
                        color: "#1B5E20",
                        iconUrl:
                            "https://res.cloudinary.com/deljcbcvu/image/upload/v1768339123/products/nn8po2g34ud2irlriuxl.png"
                    )
                    GrocerySubCategoryCard(
                        title: "Cooking Corner",
                        discount: "Up to 40% off",
                        color: "#1B5E20",
                        iconUrl:
                            "https://res.cloudinary.com/deljcbcvu/image/upload/v1768339216/products/kxl5ejlw3jyuxicbdimo.png"
                    )
                }
            }
            .padding(20)
        }
    }
}

struct GrocerySubCategoryCard: View {
    let title: String
    let discount: String
    let color: String
    let iconUrl: String

    var body: some View {
        VStack(spacing: 8) {
            Text(discount)
                .font(.system(size: 9, weight: .bold))
                .foregroundColor(.white)
                .padding(.horizontal, 8)
                .padding(.vertical, 3)
                .background(Color(hex: color))
                .cornerRadius(4)

            Text(title)
                .font(.system(size: 11, weight: .semibold))
                .foregroundColor(Color(hex: "#333"))
                .multilineTextAlignment(.center)

            // Icon
            AsyncImage(url: URL(string: iconUrl)) { image in
                image.resizable()
                    .aspectRatio(contentMode: .fit)
            } placeholder: {
                Circle()
                    .fill(Color(hex: "#F3F4F6"))
            }
            .frame(width: 50, height: 50)
        }
        .padding(10)
        .background(Color.white)
        .cornerRadius(12)
        .frame(maxWidth: .infinity)
        .shadow(color: Color.black.opacity(0.1), radius: 2, x: 0, y: 1)
    }
}

struct GrocerySectionPlaceholder: View {
    let title: String
    let color: Color

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text(title)
                .font(.system(size: 18, weight: .bold))
                .foregroundColor(Color(hex: "#111827"))
                .padding(.horizontal, 16)

            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 12) {
                    ForEach(0..<4) { _ in
                        RoundedRectangle(cornerRadius: 12)
                            .fill(Color.white)
                            .frame(width: 140, height: 180)
                            .overlay(
                                VStack {
                                    Rectangle().fill(color.opacity(0.1)).frame(height: 100)
                                    Spacer()
                                    Text("Product")
                                        .font(.system(size: 14, weight: .medium))
                                        .foregroundColor(Color(hex: "#1F2937"))
                                    Text("₹99")
                                        .font(.system(size: 14, weight: .bold))
                                        .foregroundColor(Color(hex: "#111827"))
                                    Spacer()
                                }
                                .cornerRadius(12)
                            )
                            .shadow(color: Color.black.opacity(0.05), radius: 2, x: 0, y: 1)
                    }
                }
                .padding(.horizontal, 16)
            }
        }
    }
}

File: apps/Applications/android/app/src/main/java/com/ecommerceearn/app/ui/pages/GroceryPageView.kt
package com.ecommerceearn.app.ui.pages

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Home
import androidx.compose.material.icons.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.Mic
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.viewmodel.compose.viewModel
import coil.compose.AsyncImage
import com.ecommerceearn.app.ui.components.SDUIPage
import com.ecommerceearn.app.ui.home.TopCategoryBoxesView
import com.ecommerceearn.app.ui.home.TabType
import com.ecommerceearn.app.ui.home.LocationViewModel
import com.ecommerceearn.app.data.manager.NavigationManager
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import kotlinx.coroutines.delay

@androidx.compose.foundation.ExperimentalFoundationApi
@Composable
fun GroceryPageView(
    onTabSelected: (TabType) -> Unit = {}
) {
    var searchText by remember { mutableStateOf("") }
    var isSearching by remember { mutableStateOf(false) }
    var typingText by remember { mutableStateOf("") }
    val fullText = "Lowest price..."
    var isDeleting by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) {
        while (true) {
            if (!isDeleting) {
                if (typingText.length < fullText.length) {
                    typingText = fullText.substring(0, typingText.length + 1)
                    delay(150)
                } else {
                    delay(1500)
                    isDeleting = true
                }
            } else {
                if (typingText.isNotEmpty()) {
                    typingText = typingText.substring(0, typingText.length - 1)
                    delay(100)
                } else {
                    isDeleting = false
                    delay(500)
                }
            }
        }
    }

    Box(modifier = Modifier.fillMaxSize()) {
        // Gradient Background
        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(
                    Brush.verticalGradient(
                        colors = listOf(
                            Color(0xFFFFF8E7),
                            Color(0xFFFFFDF5),
                            Color.White
                        )
                    )
                )
        )

        LazyColumn(
            modifier = Modifier.fillMaxSize(),
            contentPadding = PaddingValues(bottom = 100.dp)
        ) {
            item {
                GroceryStaticHeader(onTabSelected = onTabSelected)
            }

            stickyHeader {
                GroceryStickyHeader(
                    text = searchText,
                    onSearchTap = { isSearching = true }
                )
            }

            // Hero Banner with typing animation
            item {
                GroceryHeroBanner(typingText = typingText)
            }

            // SDUI Content
            item {
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(top = 16.dp)
                ) {
                    SDUIPage(slug = "grocery")
                }
            }
        }

        // Fixed Top padding for status bar matching iOS approach
        Box(modifier = Modifier.fillMaxWidth().statusBarsPadding().background(Color(0xFFFFF8E7)))

        // Grocery Product Detail Overlay
        val groceryProductId by NavigationManager.groceryProductId.collectAsState()
        groceryProductId?.let { productId ->
            GroceryProductDetailView(
                productId = productId,
                onBack = { NavigationManager.dismissGroceryProduct() }
            )
        }

        if (isSearching) {
            com.ecommerceearn.app.ui.pages.GlobalSearchView(
                onDismiss = { isSearching = false },
                categoryId = null
            )
        }
    }
}

@Composable
fun GroceryStaticHeader(
    locationViewModel: LocationViewModel = viewModel(),
    onTabSelected: (TabType) -> Unit
) {
    val locationState by locationViewModel.locationState.collectAsState()
    val sdfDay = SimpleDateFormat("d", Locale.getDefault())
    val sdfMonth = SimpleDateFormat("MMM", Locale.getDefault())
    val sdfDayName = SimpleDateFormat("EEEE", Locale.getDefault())
    val now = Date()

    Column(modifier = Modifier.fillMaxWidth().statusBarsPadding()) {
        TopCategoryBoxesView(
            activeTab = TabType.Grocery,
            onTabSelected = onTabSelected
        )

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 12.dp, vertical = 8.dp)
                .background(Color(0xFFFFF8E7)),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Row(
                modifier = Modifier
                    .weight(1f)
                    .clickable { NavigationManager.navigate("locationPicker") },
                verticalAlignment = Alignment.CenterVertically
            ) {
                Icon(Icons.Default.Home, contentDescription = null, tint = Color(0xFF8B6914), modifier = Modifier.size(16.dp))
                Spacer(modifier = Modifier.width(8.dp))
                Text(
                    "HOME",
                    fontSize = 11.sp,
                    fontWeight = FontWeight.Bold,
                    color = Color(0xFF8B6914)
                )
                Spacer(modifier = Modifier.width(4.dp))
                Text(
                    if (locationState.fullAddress.isEmpty()) "Select your location" else locationState.areaName,
                    fontSize = 12.sp,
                    color = Color(0xFF6B5720),
                    maxLines = 1,
                    modifier = Modifier.weight(1f, fill = false)
                )
                Icon(Icons.Default.KeyboardArrowRight, contentDescription = null, tint = Color(0xFF8B6914), modifier = Modifier.size(14.dp))
            }

            Row(
                modifier = Modifier
                    .background(Color(0xFFFF9800), RoundedCornerShape(8.dp))
                    .padding(horizontal = 12.dp, vertical = 6.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(sdfDay.format(now), fontSize = 12.sp, fontWeight = FontWeight.Bold, color = Color.White)
                Spacer(modifier = Modifier.width(2.dp))
                Text(sdfMonth.format(now), fontSize = 12.sp, fontWeight = FontWeight.Bold, color = Color.White)
                Spacer(modifier = Modifier.width(4.dp))
                Text(sdfDayName.format(now), fontSize = 9.sp, color = Color.White.copy(alpha = 0.9f))
            }
        }
    }
}

@Composable
fun GroceryStickyHeader(text: String, onSearchTap: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(Color(0xFFFFF8E7))
            .padding(horizontal = 16.dp, vertical = 8.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Row(
            modifier = Modifier
                .weight(1f)
                .background(Color.White, RoundedCornerShape(10.dp))
                .clickable { onSearchTap() }
                .padding(horizontal = 14.dp, vertical = 12.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(Icons.Default.Search, contentDescription = null, tint = Color(0xFF9CA3AF), modifier = Modifier.size(18.dp))
            Spacer(modifier = Modifier.width(8.dp))
            Text("Search for atta, dal, oil...", fontSize = 14.sp, color = Color(0xFF6B7280))
        }

        Box(
            modifier = Modifier
                .size(46.dp)
                .background(Color.White, RoundedCornerShape(10.dp))
                .clickable { onSearchTap() },
            contentAlignment = Alignment.Center
        ) {
            Icon(Icons.Default.Mic, contentDescription = null, tint = Color(0xFFFF9800))
        }
    }
}

@Composable
fun GroceryHeroBanner(typingText: String) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 8.dp)
            .height(260.dp)
            .clip(RoundedCornerShape(16.dp))
    ) {
        // Background Image
        AsyncImage(
            model = "https://res.cloudinary.com/deljcbcvu/image/upload/v1768337045/Grocery_Offer_Backgroung_krgtp0.jpg",
            contentDescription = null,
            contentScale = ContentScale.Crop,
            modifier = Modifier.fillMaxSize()
        )

        // Darker overlay for text readability
        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(Color.Black.copy(alpha = 0.15f))
        )

        Column(
            modifier = Modifier.fillMaxSize().padding(20.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            // Typing animation text
            Box(
                modifier = Modifier.height(60.dp),
                contentAlignment = Alignment.Center
            ) {
                Text(
                    text = "$typingText|",
                    fontSize = 34.sp,
                    fontWeight = FontWeight.Bold,
                    color = Color.White,
                    style = androidx.compose.ui.text.TextStyle(
                        shadow = androidx.compose.ui.graphics.Shadow(
                            color = Color.Black.copy(alpha = 0.4f),
                            offset = androidx.compose.ui.geometry.Offset(1f, 1f),
                            blurRadius = 4f
                        )
                    )
                )
            }

            Spacer(modifier = Modifier.weight(1f))

            // Subcategory Cards Row
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(10.dp)
            ) {
                GrocerySubCategoryCard(
                    title = "Winter Essential",
                    discount = "Up to 70% off",
                    color = Color(0xFF1B5E20),
                    iconUrl = "https://res.cloudinary.com/deljcbcvu/image/upload/v1768339003/products/z9nxbc93hgfkkogpflbk.png",
                    modifier = Modifier.weight(1f)
                )
                GrocerySubCategoryCard(
                    title = "Snack & Sip",
                    discount = "Up to 50% off",
                    color = Color(0xFF1B5E20),
                    iconUrl = "https://res.cloudinary.com/deljcbcvu/image/upload/v1768339123/products/nn8po2g34ud2irlriuxl.png",
                    modifier = Modifier.weight(1f)
                )
                GrocerySubCategoryCard(
                    title = "Cooking Corner",
                    discount = "Up to 40% off",
                    color = Color(0xFF1B5E20),
                    iconUrl = "https://res.cloudinary.com/deljcbcvu/image/upload/v1768339216/products/kxl5ejlw3jyuxicbdimo.png",
                    modifier = Modifier.weight(1f)
                )
            }
        }
    }
}

@Composable
fun GrocerySubCategoryCard(
    title: String,
    discount: String,
    color: Color,
    iconUrl: String,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier
            .background(Color.White, RoundedCornerShape(12.dp))
            .padding(10.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Box(
            modifier = Modifier
                .background(color, RoundedCornerShape(4.dp))
                .padding(horizontal = 8.dp, vertical = 3.dp)
        ) {
            Text(discount, fontSize = 9.sp, fontWeight = FontWeight.Bold, color = Color.White)
        }

        Spacer(modifier = Modifier.height(4.dp))

        Text(
            text = title,
            fontSize = 11.sp,
            fontWeight = FontWeight.SemiBold,
            color = Color(0xFF333333),
            textAlign = androidx.compose.ui.text.style.TextAlign.Center,
            maxLines = 2
        )

        Spacer(modifier = Modifier.height(6.dp))

        AsyncImage(
            model = iconUrl,
            contentDescription = null,
            contentScale = ContentScale.Fit,
            modifier = Modifier.size(50.dp)
        )
    }
}

